libpromises/dbm_lmdb.c

```diff
@@ -827,6 +827,66 @@ bool DBPrivWrite(
     return (rc == MDB_SUCCESS);
 }
 
+bool DBPrivOverwrite(DBPriv *db, const char *key, int key_size, const void *value, size_t value_size,
+                     OverwriteCondition Condition, void *data)
+{
+    assert(db != NULL);
+    assert(key_size >= 0);
+    DBTxn *txn;
+    int rc = GetWriteTransaction(db, &txn);
+
+    if (rc != MDB_SUCCESS)
+    {
+        return false;
+    }
+
+    assert(txn != NULL);
+    assert(!txn->cursor_open);
+
+    MDB_val mkey, orig_data;
+    mkey.mv_data = (void *) key;
+    mkey.mv_size = key_size;
+    rc = mdb_get(txn->txn, db->dbi, &mkey, &orig_data);
+    CheckLMDBCorrupted(rc, db->env);
+    if ((rc != MDB_SUCCESS) && (rc != MDB_NOTFOUND))
+    {
+        Log(LOG_LEVEL_ERR, "Could not read database entry from '%s': %s",
+            (char *) mdb_env_get_userctx(db->env), mdb_strerror(rc));
+        AbortTransaction(db);
+        return false;
+    }
+
+    void *cur_val = (rc == MDB_SUCCESS ? orig_data.mv_data : NULL);
+    size_t cur_val_size = (rc == MDB_SUCCESS ? orig_data.mv_size : 0);
+    if (Condition != NULL)
+    {
+        void *cur_val_copy = xmalloc(cur_val_size);
+        memcpy(cur_val_copy, cur_val, cur_val_size);
+        if (!Condition(cur_val_copy, cur_val_size, data))
```

Comment:
Condition is called with `size` 0 but non-null `cur_val_copy`. This fails an assertion inside the `NoOrObsoleteLock`.